[dnf5] Transaction (history)

dmach opened the pull request.

Commit:
transaction: Move files to the new location

Commit:
transaction: Fix compilation issues

Commit:
transaction: Move tests to the new location

Commit:
transaction: Fix compilation issues in tests

Commit:
transaction: Rewrite TransactionItemReason

Commit:
transaction: Rewrite TransactionItemAction

Commit:
transaction: Merge private Transaction into the public class

Commit:
transaction: Move Repo from private namespace

Commit:
transaction: Move TransformerTransaction from private namespace

Commit:
transaction: Move classes to libdnf::transaction namespace

Commit:
transaction: Rewrite Transaction class

Commit:
transaction: Merge TransactionItemBase into TransactionItem

Commit:
transaction: Rewrite TransactionItem

Commit:
transaction: Split 'Types.hpp' into individual headers

Commit:
sqlite3: Make last_insert_rowid() part of Statement public API

It is useful when passing a Statement/Query argument
and the function needs to read the inserted rowid.
Until now it wasn't possible without passing the database connection as well.

Commit:
transaction: Move RPMItem low-level db code to separate files

Commit:
transaction: Move Item low-level db code to separate files

Commit:
transaction: Move Transaction low-level db code to separate files

Commit:
transaction: Remove unused operators from TransactionItem

Commit:
transaction: Remove get*Item() methods from TransactionItem

Commit:
transaction: Remove SQLite3 from constructors, reduce shared_ptr usage

Commit:
transaction: Rewrite CompsGroupItem

Commit:
transaction: Rewrite CompsEnvironmentItem

Commit:
transaction: Rewrite RPMItem

Commit:
transaction: Drop Repo class, handle repoid in TransactionItem

Commit:
transaction: Move more Package low-level db code to separate files

Commit:
transaction: Change Package.epoch type from int to string

Make it consistent with Nevra and rpm::Package objects

Commit:
transaction: Rewrite Transaction.softwarePerformedWith to runtime_packages

The new method names better describe packages that are used as runtime used
to perform the transaction.
Also simplify the API by passing NEVRAs instead of Package objects.

Commit:
transaction: Remove Package.getNEVRA()

Commit:
transaction: Remove unused nevraToPackage() from tests

Commit:
transaction: Move Transaction.console_output low-level db code to separate files

Commit:
transaction: Move CompsGroupPackage low-level db code to separate files

Commit:
transaction: Move CompsGroup low-level db code to separate files

Commit:
transaction: Move CompsEnvironmentGroup low-level db code to separate files

Commit:
transaction: Move CompsEnvironment low-level db code to separate files

Commit:
transaction: Remove PackagePtr

Commit:
transaction: Refactor TransactionItem inheritance, keep each type in its own vector.

Commit:
sqlite3: Make get_db() part of Statement public API

It is useful when passing a Statement/Query argument
and the function needs to to use the underlying SQLite3 object.

Commit:
transaction: Implement TransactionSack

Commit:
sack: Implement protected get_data_guard() method

It is required for lazy loading of items in Query.

Commit:
weak_ptr: Implement operator*()

dmach commented:
The PR is good enough to be reviewed and hopefully integrated with Goal and RPM transaction.
There's still a lot of work to do, but I'd create new (smaller) PRs later on.

Commit:
transaction: Make header files public

j-mracek commented on include/libdnf/rpm/nevra.hpp:
```diff
@@ -149,6 +149,39 @@ inline void Nevra::set_arch(std::string && arch) {
+    if (epoch.empty()) {
+        epoch = "0";
+    }
+    return obj.get_name() + "-" + epoch + ":" + obj.get_version() + "-" + obj.get_release() + "." + obj.get_arch();
```
What about to use std::string::reserce() and append? First calculate  required size, then reserve memory, then append() strings?

dmach replied:
That sounds like over-engineering to me, I prefer to keep the code simple.
I wanted to use std::ostringstream, but it was way slower.
I ended up with this code and @jrohel seemed to be fine with it when we discussed it over chat.

dmach replied:
Ok, changed that according to your suggestion. The performance increase is up to 25%.
This is separated to https://github.com/rpm-software-management/libdnf/pull/1042

jrohel commented on include/libdnf/utils/weak_ptr.hpp:
```diff
@@ -216,6 +216,7 @@ struct WeakPtr {
     /// Checks if the other WeakPtr instance has the same WeakPtrGuard.
     bool has_same_guard(const WeakPtr & other) const noexcept { return guard == other.guard; }
 
+    TPtr * operator*() const { return get(); }
```
Operator `*` is used for dereferencing pointer. For getting pointer is `get()`.
Try:
`TPtr & operator*() const { return *get(); }`

Then you replace `CPPUNIT_ASSERT(**first > **second);` by `CPPUNIT_ASSERT(*first > *second);` in unittests

dmach replied:
fixed

jrohel commented on libdnf/transaction/private/Transaction.cpp:
```diff
@@ -55,7 +58,7 @@ swdb_private::Transaction::finish(TransactionState state)
         if (i->getState() == TransactionItemState::UNKNOWN) {
             throw std::runtime_error(
-                tfm::format(_("TransactionItem state is not set: %s"), i->getItem()->toStr()));
+                fmt::format(_("TransactionItem state is not set: %s"), i->getItem()->toStr()));
```
tinyformat (`tfm`) format string conforms to C `printf` format string -> `%s`
`fmt::format()` use Python-like formatting -> `{}`

dmach replied:
fixed

j-mracek commented on include/libdnf/transaction/comps_environment.hpp:
```diff
@@ -0,0 +1,107 @@
+    explicit CompsEnvironment(Transaction & trans);
+
+    const std::string & get_environment_id() const noexcept { return environment_id; }
+    void set_environment_id(const std::string & value) { environment_id = value; }
```
Why there is a setter in public API? Goal can use a private function as a friend.

j-mracek replied:
The same question is applicable for all setters.

dmach replied:
I'm not sure if the setters can be turned protected or private now.
They're needed in the tests and until the Goal can populate Transaction incl. transaction items,
I'm not sure if we can hide them.

j-mracek commented on include/libdnf/transaction/comps_environment.hpp:
```diff
@@ -0,0 +1,107 @@
+    /// Get list of groups associated with the environment.
+    const std::vector<std::unique_ptr<CompsEnvironmentGroup>> & get_groups() { return groups; }
+
+    //static std::vector< TransactionItemPtr > getTransactionItemsByPattern(
```
Please add TODO with description why it remains here.

dmach replied:
added

j-mracek commented on include/libdnf/transaction/comps_environment.hpp:
```diff
@@ -0,0 +1,107 @@
+    friend class CompsEnvironmentGroup;
+    std::string environment_id;
+    std::string name;
+    std::string translated_name;
```
In case of translated name, what about to store also a lang?

dmach replied:
It's not possible with the current schema. We could add TODO and eventually implement this RFE later.

j-mracek replied:
Please could you add TODO, anyway it is low priority problem with a minimal impact.

j-mracek commented on include/libdnf/transaction/comps_environment.hpp:
```diff
@@ -0,0 +1,107 @@
+    std::string name;
+    std::string translated_name;
+    CompsPackageType package_types = CompsPackageType::DEFAULT;
+    std::vector<std::unique_ptr<CompsEnvironmentGroup>> groups;
```
Does an instance of CompsEnvironment contain only one type of CompsEnvironmentGroup?
Why `CompsEnvironmentGroup` and not `CompsGroup`?

dmach replied:
It's for the same reason why `CompsGroup` doesn't contain `rpm::Package`.
Comps are weak and we store all information about groups incl. their packages that are not installed at the time of the transaction.

j-mracek replied:
Thanks

j-mracek commented on include/libdnf/transaction/comps_environment.hpp:
```diff
@@ -0,0 +1,107 @@
+public:
+    explicit CompsEnvironmentGroup(CompsEnvironment & environment);
+
+    int64_t get_id() const noexcept { return id; }
```
What represents ID? Please could you add a description?

j-mracek commented on include/libdnf/transaction/comps_environment.hpp:
```diff
@@ -0,0 +1,107 @@
+};
+
+
+class CompsEnvironmentGroup {
```
Please how the class will be used? For SWDB, for metadata representative or something else. Please could you add a description?

j-mracek commented on include/libdnf/transaction/comps_environment.hpp:
```diff
@@ -0,0 +1,107 @@
+    const CompsEnvironment & get_environment() const noexcept { return environment; }
+
+    const std::string & get_group_id() const noexcept { return group_id; }
+    void set_group_id(const std::string & value) { group_id = value; }
```
I don't think that setters should be public.

j-mracek commented on include/libdnf/transaction/comps_group.hpp:
```diff
@@ -0,0 +1,115 @@
+    DEFAULT = 1 << 1,      // installed by default, but can be unchecked in the UI
+    MANDATORY = 1 << 2,    // always installed
+    OPTIONAL = 1 << 3      // not installed by default, but can be checked in the UI
+};
```
Not sure whether the description has the right effect. Even mandatory doesn't mean must be installed.

dmach replied:
I've added a comment explaining how comps work.

j-mracek commented on include/libdnf/transaction/comps_group.hpp:
```diff
@@ -0,0 +1,115 @@
+    friend class CompsGroupPackage;
+    std::string group_id;
+    std::string name;
+    std::string translated_name;
```
What about to store LANG?

dmach replied:
It's not possible with the current schema. We could add TODO and eventually implement this RFE later.

j-mracek replied:
Translated information is available for the case that user list/remove installed groups without available repositories. Then the question is how it should be handled when lang is different from stored translated name?

j-mracek commented on include/libdnf/transaction/comps_group.hpp:
```diff
@@ -0,0 +1,115 @@
+
+class CompsGroupPackage {
+public:
+    explicit CompsGroupPackage(CompsGroup & group);
```
Why constructor doesn't require CompsPackageType?

dmach replied:
Because it also doesn't require the other attributes.
An empty object is created and all the attributes are set after.

j-mracek commented on include/libdnf/transaction/rpm_package.hpp:
```diff
@@ -0,0 +1,81 @@
+    const std::string & get_arch() const noexcept { return arch; }
+    void set_arch(const std::string & value) { arch = value; }
+
+    /*
```
Please could you add TODO with short description you keep the code?

dmach replied:
added

j-mracek commented on include/libdnf/transaction/rpm_package.hpp:
```diff
@@ -0,0 +1,81 @@
+    explicit Package(Transaction & trans);
+
+    const std::string & get_name() const noexcept { return name; }
+    void set_name(const std::string & value) { name = value; }
```
Do we realy want to allow modification of name, epoch, ... in TSI?

j-mracek commented on include/libdnf/transaction/rpm_package.hpp:
```diff
@@ -0,0 +1,81 @@
+        libdnf::utils::SQLite3 & conn, const std::vector<std::string> & patterns);
+    static TransactionItemReason resolveTransactionItemReason(
+        libdnf::utils::SQLite3 & conn, const std::string & name, const std::string & arch, int64_t maxTransactionId);
+    */
```
I suggest that TODO about get/set_reason would be nice. In this case reason setter could be public.

dmach replied:
I can do that, but I'm already finishing another PR with the reasons.

jrohel commented on include/libdnf/transaction/transaction.hpp:
```diff
@@ -0,0 +1,221 @@
+    /// The method must be called only on a started transaction
+    void add_console_output_line(int file_descriptor, const std::string & line);
+
+    void begin();
```
I'm thinking of different name, because `begin()` is usually used for iterators. And I don't want to close doors for implementing standard iterators. May it be `start()`? Or something else?

j-mracek commented on include/libdnf/transaction/transaction_item.hpp:
```diff
@@ -0,0 +1,105 @@
+
+protected:
+    int64_t id = 0;
+    Action action = Action::INSTALL;
```
To setup install as a default action doesn't sound right to me.

j-mracek replied:
Constructor with minimal arguments is one thing, but new object with incorrect default data is not a good approach.

j-mracek commented on include/libdnf/transaction/transaction_item.hpp:
```diff
@@ -0,0 +1,105 @@
+    Action action = Action::INSTALL;
+    Reason reason = Reason::UNKNOWN;
+    State state = State::UNKNOWN;
+    std::string repoid;
```
Will we store repoid also for modules?

j-mracek commented on include/libdnf/transaction/transaction_item_type.hpp:
```diff
@@ -0,0 +1,39 @@
+    UNKNOWN = 0, // default type, must be changed before save
+    RPM = 1,
+    GROUP = 2,
+    ENVIRONMENT = 3
```
Modules?

j-mracek commented on libdnf/transaction/MergedTransaction.cpp:
```diff
@@ -19,37 +20,46 @@
+{
+}
+*/
+
```
Why you keep the comment?

j-mracek commented on libdnf/transaction/MergedTransaction.cpp:
```diff
@@ -209,24 +220,68 @@ MergedTransaction::getItems()
+    return env->get_environment_id();
+}
+
+/*
```
I am sorry but I don't understand why the note is here? It was a part of some function but now it is only left over.

j-mracek commented on libdnf/transaction/MergedTransaction.cpp:
```diff
@@ -235,50 +290,48 @@ getItemIdentifier(ItemPtr item)
-    auto firstItem = previousItemPair.first->getItem();
-    auto secondItem = mTransItem->getItem();
+void MergedTransaction::resolveRPMDifference(ItemPair & previousItemPair, TransactionItem * mTransItem) {
+    auto * firstItem = previousItemPair.first;
```
CamelCase? Not `first_item`?

j-mracek commented on libdnf/transaction/MergedTransaction.hpp:
```diff
@@ -0,0 +1,87 @@
+#include <set>
+#include <string>
+#include <map>
+#include <vector>
```
Please could you properly order imports?
Please could you run clang on all changed files?

j-mracek commented on libdnf/transaction/Swdb.cpp:
```diff
@@ -19,33 +21,35 @@
  */
 
 #include <cstdio>
+#include <filesystem>
```
The order of includes is incorrect.

j-mracek commented on libdnf/transaction/Swdb.hpp:
```diff
@@ -27,37 +29,34 @@
-#include "../sack/packageset.hpp"
-#include "../utils/sqlite3/Sqlite3.hpp"
+//#include "../hy-types.h"
+//#include "../sack/query.hpp"
```
Please remove unused includes

j-mracek replied:
Please could you order includes properly. Clang will help.

j-mracek commented on libdnf/transaction/Swdb.hpp:
```diff
@@ -87,16 +95,18 @@ struct Swdb {
                                                           int64_t maxTransactionId);
     const std::string getRPMRepo(const std::string &nevra);
-    TransactionItemPtr getRPMTransactionItem(const std::string &nevra);
+    //TransactionItemPtr getRPMTransactionItem(const std::string &nevra);
```
Please why did you create a comment?

j-mracek commented on libdnf/transaction/Swdb.hpp:
```diff
@@ -111,25 +121,30 @@ struct Swdb {
 
     /**
     * @brief Remove packages from PackageSet that were installed as Dependency or WEAK_DEPENDENCY
     */
```
Please use `///` for method description

j-mracek commented on libdnf/transaction/Transformer.cpp:
```diff
@@ -453,89 +457,101 @@ Transformer::transformRPMItems(SQLite3Ptr swdb,
     }
 
-    compsGroup->save();
+    //compsGroup->save();
```
Please remove it or add todo.

j-mracek commented on libdnf/transaction/Transformer.cpp:
```diff
@@ -453,89 +457,101 @@ Transformer::transformRPMItems(SQLite3Ptr swdb,
 /**
  * Construct CompsEnvironmentItem object from JSON
  * \param env environment json object
  */
```
Now we use `//` instead `/*`

j-mracek commented on libdnf/transaction/Transformer.cpp:
```diff
@@ -453,89 +457,101 @@ Transformer::transformRPMItems(SQLite3Ptr swdb,
+            auto & grp = compsEnv->new_group();
+            grp.set_group_id(key);
+            grp.set_installed(false);
+            grp.set_group_type(CompsPackageType::MANDATORY);
```
Let me think. In case that constructor of EnvGroup will be private and also method compsEnv->new_group(), then we can keep the original interface `compsEnv->add_group(key, false, CompsPackageType::MANDATORY);`. A also suggest to rename `compsEnv` to `comps_env`.

j-mracek commented on libdnf/transaction/Transformer.cpp:
```diff
@@ -453,89 +457,101 @@ Transformer::transformRPMItems(SQLite3Ptr swdb,
     }
 
-    compsEnv->save();
+    //compsEnv->save();
```
Please add todo why you keep the note and what's the plan.

j-mracek commented on libdnf/transaction/Transformer.hpp:
```diff
@@ -25,16 +27,16 @@
-#include "private/Transaction.hpp"
-#include "private/TransformerTransaction.hpp"
-#include "TransactionItem.hpp"
+#include "comps_environment.hpp"
```
Please correct the order of includes.

j-mracek commented on libdnf/transaction/TransformerTransaction.hpp:
```diff
@@ -21,24 +21,29 @@
-namespace libdnf {
+#include "transaction.hpp"
+
+#include "libdnf/transaction/db/trans_with.hpp"
```
I suggest that includes are in not correct order, but not 100% sure.

j-mracek commented on libdnf/transaction/comps_environment.cpp:
```diff
@@ -0,0 +1,83 @@
+CompsEnvironment::CompsEnvironment(Transaction & trans) : TransactionItem::TransactionItem(trans, Type::ENVIRONMENT) {}
+
+
+/*
```
Please could you add TODO?

j-mracek commented on libdnf/transaction/comps_environment.cpp:
```diff
@@ -0,0 +1,83 @@
+CompsEnvironmentGroup & CompsEnvironment::new_group() {
+    auto grp = std::make_unique<CompsEnvironmentGroup>(*this);
+    // TODO(dmach): following lines are not thread-safe
+    groups.push_back(std::move(grp));
```
What about:
```
groups.emplace_back(std::make_unique<CompsEnvironmentGroup>(*this));
````

j-mracek commented on libdnf/transaction/comps_group.cpp:
```diff
@@ -0,0 +1,84 @@
+CompsGroup::CompsGroup(Transaction & trans) : TransactionItem::TransactionItem(trans, Type::GROUP) {}
+
+
+/*
```
Please add todo.

j-mracek commented on libdnf/transaction/comps_group.cpp:
```diff
@@ -0,0 +1,84 @@
+CompsGroupPackage & CompsGroup::new_package() {
+    auto pkg = std::make_unique<CompsGroupPackage>(*this);
+    // TODO(dmach): following lines are not thread-safe
+    packages.push_back(std::move(pkg));
```
What about to use `emplace_back(std::make_unique<CompsGroupPackage>(*this))`. It will skip all move steps and pointer will be created directly in vector.

dmach replied:
The latest changes (that are still not pushed) where I changed the constructor to protected made it even worse.
You can't use std::make_unique, because unique_ptr is not friend with the CompsGroupPackage class, but CompsGroup is.
In such case you need to create a pointer and turn it into an unique_ptr later on.

j-mracek commented on libdnf/transaction/db/comps_environment.cpp:
```diff
@@ -0,0 +1,137 @@
+
+#include "libdnf/transaction/comps_group.hpp"
+#include "libdnf/transaction/transaction.hpp"
+#include "libdnf/transaction/transaction_item.hpp"
```
Not sure if order is correct. Clang could know better.

j-mracek commented on libdnf/transaction/db/comps_environment.hpp:
```diff
@@ -0,0 +1,57 @@
+
+
+/// Return a vector of CompsEnvironment objects with comps environments in a transaction
+std::vector<std::unique_ptr<CompsEnvironment>> get_transaction_comps_environments(libdnf::utils::SQLite3 & conn, Transaction & trans);
```
Why those 4 functions are in a separate module? Who owns libdnf::utils::SQLite3 & conn?

j-mracek commented on libdnf/transaction/db/comps_group.hpp:
```diff
@@ -0,0 +1,57 @@
+
+
+/// Return a vector of CompsGroup objects with comps groups in a transaction
+std::vector<std::unique_ptr<CompsGroup>> get_transaction_comps_groups(libdnf::utils::SQLite3 & conn, Transaction & trans);
```
Why do you prefer to have it in a separate module?

j-mracek commented on libdnf/transaction/db/comps_group_package.cpp:
```diff
@@ -0,0 +1,102 @@
+
+#include "comps_group_package.hpp"
+
+#include "libdnf/transaction/transaction.hpp"
```
Please could you reorder includes?

dmach replied:
clang-format keeps it this way actually.
If the include file name is the same as the current file (minus the suffix), it's placed on top.

j-mracek commented on libdnf/transaction/db/item.cpp:
```diff
@@ -0,0 +1,53 @@
+        throw std::runtime_error("");
+    }
+    return query.last_insert_rowid();
+}
```
Do we really want to split the code into so small pieces?

dmach replied:
For the moment - yes.
I definitely want to make additional changes to the code, but I'd prefer that to be in the next iteration without touching this PR.

Commit:
transaction: Apply clang-format

j-mracek commented on libdnf/transaction/db/rpm.hpp:
```diff
@@ -0,0 +1,83 @@
+namespace libdnf::transaction {
+
+
+class Package;
```
I suggest that Package class is not rpm package, but swdb package. We have to ensure that both Package classes are exportable into multiple languages.

dmach replied:
I don't understand the suggestion. Each Package lives in another module (libdnf.rpm.Package vs libdnf.transaction.Package) if that's what you're looking for.

j-mracek replied:
Ok, then it should be fine when it is in another module.

j-mracek commented on libdnf/transaction/db/trans_item.cpp:
```diff
@@ -0,0 +1,91 @@
+
+#include "trans_item.hpp"
+
+#include "repo.hpp"
```
Please could you reorder includes?

j-mracek commented on libdnf/transaction/db/trans_with.cpp:
```diff
@@ -0,0 +1,126 @@
+
+    for (auto & i : trans.get_runtime_packages()) {
+        libdnf::rpm::Nevra nevra;
+        nevra.parse(i, libdnf::rpm::Nevra::Form::NEVRA);
```
Please don't use `i` as name for nevra string variable. I see `i` as an index. Also I don't like a construct when you store NEVRA in transaction in progress instead of libdnf::rpm::Package. Then the parse step will be not required. Or did I miss something?

dmach replied:
Changed to 'ti' which stands for TransactionItem and is consistent with the rest of the code (especially the sql queries)

j-mracek commented on libdnf/transaction/db/trans_item.cpp:
```diff
@@ -0,0 +1,91 @@
+
+
+void transaction_item_select(libdnf::utils::SQLite3::Query & query, TransactionItem & ti) {
+    ti.set_id(query.get<int64_t>("id"));
```
`query.get<int64_t>("id")` looks strange to me.

dmach replied:
Could you be more specific?
Does it mean you don't understand the code just by looking at it and it requires a comment explaining it?

j-mracek commented on libdnf/transaction/query.cpp:
```diff
@@ -0,0 +1,126 @@
+        case libdnf::sack::QueryCmp::NOT_GLOB:
+        case libdnf::sack::QueryCmp::IGLOB:
+        case libdnf::sack::QueryCmp::NOT_IGLOB:
+            // TODO(dmach): Replace with a different exception class
```
What about to replace it with a default statement?

j-mracek commented on libdnf/transaction/rpm_package.cpp:
```diff
@@ -0,0 +1,199 @@
+#include "libdnf/transaction/rpm_package.hpp"
+
+#include "libdnf/rpm/nevra.hpp"
+#include "libdnf/transaction/db/rpm.hpp"
```
Please could you reorder includes?

j-mracek commented on libdnf/transaction/rpm_package.cpp:
```diff
@@ -0,0 +1,199 @@
+ * This method doesn't care about compare package names
+ * \param other Package to compare with
+ * \return true if other package is newer (has higher version and/or epoch)
+ */
```
Incorrect format of dock string => `///`.

dmach replied:
I've removed the doc string completely.
It belongs to hpp and also such operators do not need any explanation.
There's already TODO in the code to improve it.

j-mracek commented on libdnf/transaction/rpm_package.cpp:
```diff
@@ -0,0 +1,199 @@
+
+
+
+/*
```
Please could you add TODO to describe plans for the code/comment?

dmach replied:
It's in the hpp already.

j-mracek commented on libdnf/transaction/rpm_package.cpp:
```diff
@@ -0,0 +1,199 @@
+
+    std::string bufferThis;
+    std::string bufferOther;
+    while (std::getline(versionThis, bufferThis, '.') && std::getline(versionOther, bufferOther, '.')) {
```
Why did you not use a `pool_evrcmp` functionality. This is like a reimplemented wheel with probably a different shape.

dmach replied:
This is an old code I did not touch except fixing variable / method names etc.
Not sure if pool_evrcmp() is usable, because the data is not in the pool.
On the other hand, it's a good find actually. I've added a TODO to replace the whole function with a different implementation, preferably an existing code.

j-mracek replied:
pool_evrcmp() requres pool to create tmp cache. Thanks for TODO

j-mracek commented on libdnf/transaction/rpm_package.cpp:
```diff
@@ -0,0 +1,199 @@
+    return false;
+}
+
+/*
```
Please could you add a TODO with description of future plans?

dmach replied:
I've added the TODO in the header file already.

j-mracek commented on libdnf/transaction/transaction.cpp:
```diff
@@ -0,0 +1,169 @@
+*/
+
+
+#include "libdnf/transaction/transaction.hpp"
```
Please could you reorder includes?

j-mracek commented on libdnf/transaction/transaction_item.cpp:
```diff
@@ -0,0 +1,139 @@
+*/
+
+
+#include "libdnf/transaction/transaction_item.hpp"
```
Please could you reorder includes?

j-mracek commented on libdnf/transaction/transaction_item.cpp:
```diff
@@ -0,0 +1,139 @@
+}
+
+
+/*
```
Please could you add TODO with some description why you kept the code?

dmach replied:
I've added the TODO in the header file already.

j-mracek commented on libdnf/transaction/transaction_item_reason.cpp:
```diff
@@ -0,0 +1,108 @@
+    if (lhs == rhs) {
+        return false;
+    }
+    for (auto i : order) {
```
Please could you use something more descriptive then `i` for itemReason?

dmach replied:
changed, but it's nothing exceptional that i, j, k are used as loop variables especially if the loop is not too long

j-mracek replied:
Same I can say same about p, pp in libsolv (Id). Using more descriptive names for variable increase quality of code. Thanks for the change.

j-mracek commented on test/libdnf/transaction/MergedTransactionTest.cpp:
```diff
@@ -1,96 +1,79 @@
-#include "libdnf/nevra.hpp"
-#include "libdnf/transaction/RPMItem.hpp"
+//#include "libdnf/hy-subject.h"
+#include "libdnf/rpm/nevra.hpp"
```
Please could you reorder includes? Clang could help.

j-mracek commented on test/libdnf/transaction/MergedTransactionTest.cpp:
```diff
@@ -1,96 +1,79 @@
-static libdnf::swdb_private::TransactionPtr
-initTransFirst(SQLite3Ptr conn)
+static TransactionPtr
+initTransFirst(libdnf::utils::SQLite3 & conn)
```
What about `init_trans_first()`

dmach replied:
The whole file is disabled with #define SKIP macro.
The additional changes will take about a week and couple more commits.

j-mracek replied:
Then mark skip macro with TODO please with a list what have to be performed.

j-mracek commented on test/libdnf/transaction/MergedTransactionTest.cpp:
```diff
@@ -1,96 +1,79 @@
 {
-    conn = std::make_shared< SQLite3 >(":memory:");
-    Transformer::createDatabase(conn);
+    conn = new libdnf::utils::SQLite3(":memory:");
```
What about to use `unique_ptr`, then it will not require delete in tear_down()

j-mracek commented on test/libdnf/transaction/MergedTransactionTest.cpp:
```diff
@@ -324,33 +306,29 @@ MergedTransactionTest::testMergeAlterAlter()
 
-static RPMItemPtr
-nevraToRPMItem(SQLite3Ptr conn, std::string nevra)
+static std::shared_ptr<Package>
```
Why `shared_ptr`?

j-mracek replied:
But it is not a problem.

j-mracek commented on test/libdnf/transaction/MergedTransactionTest.hpp:
```diff
@@ -1,10 +1,12 @@
+#include "libdnf/utils/sqlite3/sqlite3.hpp"
 
 class MergedTransactionTest : public CppUnit::TestCase {
     CPPUNIT_TEST_SUITE(MergedTransactionTest);
```
Probably TDO would be fine here? No additional description required.

j-mracek commented on test/libdnf/transaction/MergedTransactionTest.hpp:
```diff
@@ -60,7 +62,9 @@ class MergedTransactionTest : public CppUnit::TestCase {
     void test_multilib_identity();
 private:
-    std::shared_ptr< SQLite3 > conn;
+    libdnf::utils::SQLite3 * conn;
```
What about to use `unique_ptr`. In case of raw ptr, please add {nullptr} otherwise covscan will complain.

j-mracek commented on test/libdnf/transaction/TransactionItemReasonTest.cpp:
```diff
@@ -1,40 +1,41 @@
+#ifdef SKIP
```
Why you want to mark it as 'skip'? Probably todo would be nice here.

j-mracek commented on test/libdnf/transaction/TransactionItemReasonTest.cpp:
```diff
@@ -1,40 +1,41 @@
+
 #include <cstdio>
 #include <iostream>
 #include <string>
```
Please could you reorder includes?

j-mracek commented on test/libdnf/transaction/TransformerTest.cpp:
```diff
@@ -0,0 +1,216 @@
+#ifdef SKIP
```
Plerase could you explain why it is marked as skip?

dmach replied:
Explained, added TODO.

j-mracek commented on test/libdnf/transaction/TransformerTest.cpp:
```diff
@@ -0,0 +1,216 @@
+#include <json.h>
+#include <set>
+#include <sstream>
+#include <string>
```
Please could you reorder includes?

j-mracek commented on test/libdnf/transaction/TransformerTest.cpp:
```diff
@@ -0,0 +1,216 @@
+TransformerTest::setUp()
+{
+    swdb = new libdnf::utils::SQLite3(":memory:");
+    history = new libdnf::utils::SQLite3(":memory:");
```
swdb and history could use `unique_ptr`. What do you think?

j-mracek commented on test/libdnf/transaction/TransformerTest.cpp:
```diff
@@ -0,0 +1,216 @@
+}
+
+void
+TransformerTest::testGroupTransformation()
```
It could be renamed to `test_group_transformation`

j-mracek commented on test/libdnf/transaction/TransformerTest.hpp:
```diff
@@ -29,8 +31,10 @@ class TransformerTest : public CppUnit::TestCase {
-    std::shared_ptr< SQLite3 > swdb;
-    std::shared_ptr< SQLite3 > history;
+    libdnf::utils::SQLite3 * swdb;
+    libdnf::utils::SQLite3 * history;
```
It could use `unique_ptr`

j-mracek commented on test/libdnf/transaction/test_comps_environment.hpp:
```diff
@@ -0,0 +1,41 @@
+#include "transaction_test_base.hpp"
+
+#include <cppunit/TestCase.h>
+#include <cppunit/extensions/HelperMacros.h>
```
Not sure whether the order is correct?

dmach replied:
`clang-format -i` keeps it this way

Commit:
transaction: Add docstrings

j-mracek commented on test/libdnf/transaction/test_transaction.cpp:
```diff
@@ -0,0 +1,168 @@
+    trans->set_user_id(1000);
+    trans->set_cmdline("dnf install foo");
+
+    trans->add_runtime_package("rpm-4.14.2-1.fc29.x86_64");
```
I suggest that this a weak point of implementation. It must contain libdnf::rpm::package, othewise it will be impossible to use it for rpm transaction. Connection with nevra and repoid is not unique.

dmach replied:
The goal of this PR was to improve the transaction/history db API and reduce the overall complexity.
Integration with rpm transaction is out of the scope of this PR in my opinion.

j-mracek commented on test/libdnf/transaction/test_transaction.hpp:
```diff
@@ -0,0 +1,49 @@
+
+#include "transaction_test_base.hpp"
+
+#include <cppunit/TestCase.h>
```
I am not sure whether the order is correct.

dmach replied:
`clang-format -i` keeps it this way

j-mracek commented on test/libdnf/transaction/test_workflow.cpp:
```diff
@@ -0,0 +1,148 @@
+    rpm_bash.set_release("5.fc26");
+    rpm_bash.set_arch("x86_64");
+    rpm_bash.set_repoid("base");
+    rpm_bash.set_action(TransactionItemAction::INSTALL);
```
Data are important to save it into swdb, but are useless for runtime. rpm_bash object cannot be used for rpm_transaction.

j-mracek replied:
What is the plan for the future?

dmach replied:
I take it as a note or a statement that doesn't require any change.
Or did you want to change anything?

j-mracek replied:
I think that we agreed that it will be refactored during integration with rpm transaction.

j-mracek commented on test/libdnf/transaction/test_workflow.cpp:
```diff
@@ -0,0 +1,148 @@
+
+    auto & minimal_core = comps_environment_minimal.new_group();
+    minimal_core.set_group_id("core");
+    minimal_core.set_installed(true);
```
Setting installed as true before installation is performed is not nice. Probably we have to discuss about.

dmach replied:
I've added a comment explaining why set_installed() is used.

j-mracek commented on test/libdnf/transaction/test_workflow.cpp:
```diff
@@ -0,0 +1,148 @@
+
+    for (auto & i : trans->get_comps_environments()) {
+        i->set_state(TransactionItemState::DONE);
+        // i->save();
```
Please do not use `i` as a name for variable. `tsi` is much better.

j-mracek replied:
Why save() is not performed or kept as a comment?

Commit:
transaction: Change 'begin' to 'start' in the Transaction class

The 'begin()' method might confuse API users, because it's widely used with iterators

jrohel commented:
PR brings only part of the functionality. However, it contains a lot of commits. We must go ahead. Some of pointed out problems were fixed. I will merge it. Work can continue in the next PR.
